File: Java/DesignPatterns/Command.java
import java.util.*;

public class Document
{
    String content;

    public Document()
    {
        content = "";
    }

    public void PrintDocument()
    {
        System.out.println(content);
    }

    public void AddContent(String content)
    {
        this.content += content;
    }

    public int Length()
    {
        return content.length();
    }

    public String CharAt(int index)
    {
        return String.valueOf(content.charAt(index));
    }

    public void RemoveContent(int numChars)
    {
        if(numChars < 0) return;
        if(numChars > content.length() - 1)
        {
            content = "";
            return;
        }

        //content = content.Remove(content.Length - numChars);
        content = content.substring(0,  content.length() - numChars);
        
    }
}

public interface IUndoableCommand
{
    void Execute();
    void Undo();
    void Redo();
}

public class AddContentCommand implements IUndoableCommand
{
    Document document;
    String content;

    public AddContentCommand(Document document, String content)
    {
        this.document = document;
        this.content = content;
    }

    public void Execute()
    {
        this.document.AddContent(content);
    }

    public void Undo()
    {
        this.document.RemoveContent(content.length());
    }

    public void Redo()
    {
        this.Execute();
    }
}

public class BackspaceCommand implements IUndoableCommand
{
    Document document;
    String removedChar;

    public BackspaceCommand(Document document)
    {
        this.document = document;
        removedChar = null;
    }

    public void Execute()
    {
        removedChar = document.CharAt(document.Length() - 1);
        this.document.RemoveContent(1);
    }

    public void Undo()
    {
        this.document.AddContent(this.removedChar);
    }

    public void Redo()
    {
        this.Execute();
    }
    
}

public class DocumentWriter
{
    Stack<IUndoableCommand> undoStack;
    Stack<IUndoableCommand> redoStack;

    Document document;

    public DocumentWriter(Document document)
    {
        this.document = document;
        this.undoStack = new Stack<IUndoableCommand>();
        this.redoStack = new Stack<IUndoableCommand>();
    }

    public void Write(String content)
    {
        IUndoableCommand command = new AddContentCommand(this.document, content);

        this.undoStack.push(command);
        this.redoStack.clear();

        command.Execute();
    }

    public void Backspace()
    {
        if(this.document.Length() == 0) return;

        IUndoableCommand command  = new BackspaceCommand(this.document);
        this.undoStack.push(command);
        this.redoStack.clear();

        command.Execute();
    }

    public void Undo()
    {
        if(this.undoStack.size() > 0)
        {
            IUndoableCommand command = this.undoStack.pop();
            this.redoStack.push(command);
            command.Undo();
        }
    }

    public void Redo()
    {
        if(this.redoStack.size() > 0)
        {
            IUndoableCommand command = this.redoStack.pop();
            this.undoStack.push(command);
            command.Redo();
        }
    }
}


// Document Writer Demo:
public class Solution
{
    public static void Main(String[] args)
    {
        
        Document doc = new Document();
        DocumentWriter writer = new DocumentWriter(doc);

        // Write to the document:
        writer.Write("Hello world!");

        writer.Backspace();
        doc.PrintDocument(); // Hello world
        writer.Undo();
        doc.PrintDocument(); // Hello world!
        writer.Redo();
        doc.PrintDocument(); // Hello world

        writer.Write(" Goodbye!"); 
        doc.PrintDocument(); // Hello world Goodbye!
        writer.Undo();
        doc.PrintDocument(); // Hello world
        writer.Undo();
        doc.PrintDocument(); // Hello world!

    }
}

File: TypeScript/DesignPatterns/Command.ts
// Command pattern in TypeScript

class TextDocument
{
    protected content: string;

    constructor()
    {
        this.content = "";
    }

    public printDocument()
    {
        console.log(this.content);
    }

    public addContent(content: string)
    {
        this.content += content;
    }

    public length(): number
    {
        return this.content.length;
    }

    public charAt(index: number): string
    {
        return this.content[index];
    }

    public removeContent(numChars: number)
    {
        if(numChars < 0) return;
        if(numChars > this.content.length - 1)
        {
            this.content = "";
            return;
        }

        this.content = this.content.substring(0, this.content.length - numChars);

    }
}

interface IUndoableCommand
{
    execute(): void;
    undo(): void;
    redo(): void;
}

class AddContentCommand implements IUndoableCommand
{
    document: TextDocument;
    content: string;
    
    constructor(document: TextDocument, content: string)
    {
        this.document = document;
        this.content = content;
    }

    execute() {
        this.document.addContent(this.content);
    }
    undo() {
        this.document.removeContent(this.content.length);
    }
    redo() {
        this.execute();
    }
    
}

class BackspaceCommand implements IUndoableCommand
{
    document: TextDocument;
    removedChar: string;
    
    constructor(document: TextDocument)
    {
        this.document = document;
        this.removedChar = null;
    }

    execute()
    {
        this.removedChar = this.document.charAt(this.document.length() - 1);
        this.document.removeContent(1);
    }
    undo() 
    {
        this.document.addContent(this.removedChar);
    }
    redo()
    {
        this.execute();
    }
    
}

class DocumentWriter
{
    undoStack: Array<IUndoableCommand>;
    redoStack: Array<IUndoableCommand>;

    document: TextDocument;

    constructor(document: TextDocument)
    {
        this.document = document;
        this.undoStack = new Array<IUndoableCommand>();
        this.redoStack = new Array<IUndoableCommand>();
    }

    public write(content: string)
    {
        const command = new AddContentCommand(this.document, content);
        this.undoStack.push(command);
        this.redoStack = new Array<IUndoableCommand>();

        command.execute();
    }

    public backspace()
    {
        if(this.document.length() == 0) return;
        const command = new BackspaceCommand(this.document);
        this.undoStack.push(command);
        this.redoStack = new Array<IUndoableCommand>();

        command.execute();
    }

    public undo()
    {
        if(this.undoStack.length > 0)
        {
            const command = this.undoStack.pop();
            this.redoStack.push(command);
            command.undo();
            
        }
    }

    public redo()
    {
        if(this.redoStack.length > 0)
        {
            const command = this.redoStack.pop();
            this.undoStack.push(command);
            command.redo();
        }
    }
}


class CommandSolution
{
    public execute()
    {
        let doc = new TextDocument();
        let writer = new DocumentWriter(doc);

        // Write to the document:
        writer.write("Hello world!");
        doc.printDocument(); // Hello world!
        writer.backspace();
        doc.printDocument(); // Hello world
        writer.undo();
        doc.printDocument(); // Hello world!
        writer.redo();
        doc.printDocument(); // Hello world

        writer.write(" Goodbye!"); 
        doc.printDocument(); // Hello world Goodbye!
        writer.undo();
        doc.printDocument(); // Hello world
        writer.undo();
        doc.printDocument(); // Hello world!


    }
}

const solution = new CommandSolution();
solution.execute();